docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+# ADR 054: Semver Compatible SDK Modules
+
+## Changelog
+
+* 2022-04-27: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+In order to move the Cosmos SDK to a system of decoupled semantically versioned
+modules which can be composed in different combinations (ex. staking v3 with
+bank v1 and distribution v2), we need to reassess how we organize the API surface
+of modules to avoid problems with go semantic import versioning and
+circular dependencies. This ADR explores various approaches we can take to
+addressing these issues.
+
+## Context
+
+There has been [a fair amount of desire](https://github.com/cosmos/cosmos-sdk/discussions/10162)
+in the community for semantic versioning in the SDK and there has been significant
+movement to splitting SDK modules into [standalone go modules](https://github.com/cosmos/cosmos-sdk/issues/11899).
+Both of these will ideally allow the ecosystem to move faster because we won't
+be waiting for all dependencies to update synchronously. For instance, we could
+have 3 versions of the core SDK compatible with the latest 2 releases of
+CosmWasm as well as 4 different versions of staking . This sort of setup would
+allow early adopters to aggressively integrate new versions, while allowing 
+more conservative users to be selective about which versions they're ready for.
```

Comment:
Is this not already possible today via forking?

Reply:
> @peterbourgon , today , you can't user latest Cosmos SDK with any cosmwasm. Also, you are not able to decide which updates you want to use (eg x/nft or x/group, but keep the other parts of SDK in 0.45). Of course, this ADR doesn't want to solve all possible combination, because each module will have a minimum dependencies to meet.

Of course! But this is a consequence of the design of the Cosmos SDK, and this ADR doesn't have any effect on that inter-module compatibility. You might be able to compile a binary that composed modules at versions that were previously uncompilable, but that doesn't mean those incompatible module versions would somehow become compatible, it would just move the errors from compile-time to run-time.

Reply:
The goal is to break it down, and separate tight coupling between modules whenever possible.